stan/math/opencl/kernel_generator/cast.hpp

```diff
@@ -0,0 +1,101 @@
+#ifndef STAN_MATH_OPENCL_KERNEL_GENERATOR_CAST_HPP
+#define STAN_MATH_OPENCL_KERNEL_GENERATOR_CAST_HPP
+#ifdef STAN_OPENCL
+
+#include <stan/math/prim/meta.hpp>
+#include <stan/math/opencl/matrix_cl_view.hpp>
+#include <stan/math/opencl/kernel_generator/common_return_scalar.hpp>
+#include <stan/math/opencl/kernel_generator/type_str.hpp>
+#include <stan/math/opencl/kernel_generator/name_generator.hpp>
+#include <stan/math/opencl/kernel_generator/operation_cl.hpp>
+#include <stan/math/opencl/kernel_generator/as_operation_cl.hpp>
+#include <array>
+#include <string>
+#include <type_traits>
+#include <set>
+#include <utility>
+
+namespace stan {
+namespace math {
+
+/** \addtogroup opencl_kernel_generator
+ *  @{
+ */
+
+/**
+ * Represents a typecast os scalar in kernel generator expressions.
+ * @tparam Derived derived type
+ * @tparam T type of argument
+ * @tparam Scal type of the scalar of result
+ */
+template <typename Scal, typename T>
+class cast_ : public operation_cl<cast_<Scal, T>, Scal, T> {
+ public:
+  using Scalar = Scal;
+  using base = operation_cl<cast_<Scal, T>, Scalar, T>;
+  using base::var_name_;
+
+  /**
+   * Constructor
+   * @param args argument expression(s)
+   */
+  explicit cast_(T&& arg) : base(std::forward<T>(arg)) {}
+
+  /**
+   * Generates kernel code for this expression.
+   * @param row_index_name row index variable name
+   * @param col_index_name column index variable name
+   * @param view_handled whether whether caller already handled matrix view
+   * @param var_names_arg variable names of the nested expressions
+   * @return part of kernel with code for this expression
+   */
+  inline kernel_parts generate(const std::string& row_index_name,
+                               const std::string& col_index_name,
+                               const bool view_handled,
+                               const std::string& var_name_arg) const {
+    kernel_parts res{};
+
+    res.body = type_str<Scalar>() + " " + var_name_ + " = ("
+               + type_str<Scalar>() + ")" + var_name_arg + ";\n";
```

Comment:
So this line here generates `Scalar var_name = (Scalar)input_var_name;`? If so then idt I'm understanding how this works for `char`?

Reply:
Yes. I don't understand what is the confusion with char. It works exactly the same as for any other supported scalar type.

> Side note, not necessary for this PR, but I think this is an example where it would be nice to have something in the tests or an easy scheme available from runTests.py for generating the OpenCL kernel code.

We do have reference kernels in tests, although I am not using them for operations that generate really trivial code, such  as typecasting in this PR.

Reply:
So I'm looking at 

```
auto any_n_negative = colwise_max(cast<char>(n < 0));
```

And here this is turned into 

```cpp
char var_name = (char)input_var_name;
```

Apologies if this is silly but I'm not following how that is used in an if etc. Like is it that if `input_var_name` is 0 then the `char` is 0 and the if is false?

Reply:
Yes, exactly. `input_var_name` is in this case the result of the comparison `n < 0`, which is either 0 or 1. 0 or 1 converted into char is still 0 or 1. That gets written into global memory and than copied to host. Is there still anything you are not understanding?

Reply:
Aight then I think that makes sense